torchgeo/datasets/mmearth.py

```diff
@@ -0,0 +1,505 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""MM-Earth Dataset."""
+
+import json
+import os
+from collections.abc import Callable, Sequence
+from typing import Any, ClassVar, cast
+
+import numpy as np
+import torch
+from torch import Tensor
+
+from .errors import DatasetNotFoundError
+from .geo import NonGeoDataset
+from .utils import Path, lazy_import
+
+
+class MMEarth(NonGeoDataset):
+    """MM-Earth dataset.
+
+    There are three different versions of the dataset, that vary in image size
+    and the number of tiles:
+
+    * MMEarth: 128x128 px, 1.2 M tiles, 579 GB
+    * MMEarth64: 64x64 px, 1.2 M tiles, 162 GB
+    * MMEarth100k: 128x128 px, 100 K tiles, 48 GB
+
+    The dataset consists of 12 modalities:
+
+    * Aster: elevation and slope
+    * Biome: 14 terrestrial ecosystem categories
+    * ETH Canopy Height: Canopy height and standard deviation
+    * Dynamic World: 9 landcover categories
+    * Ecoregion: 846 ecoregion categories
+    * era5: Climate reanalysis data for temperature mean, min, and max of [year, month, previous month]
+        and precipitation total of [year, month, previous month] (counted as separate modalities)
+    * ESA World Cover: 11 landcover categories
+    * Sentinel-1: VV, VH, HV, HH for ascending/descending orbit
+    * Sentinel-2: multi-spectral B1-B12 for L1C/L2A products
+    * Geolocation: cyclic encoding of latitude and longitude
+    * Date: cyclic encoding of month
+
+    Additionally, there are three masks available as modalities:
+
+    * Sentinel-2 Cloudmask: Sentinel-2 cloud mask
+    * Sentinel-2 Cloud probability: Sentinel-2 cloud probability
+    * Sentinel-2 SCL: Sentinel-2 scene classification
+
+    that are synchronized across tiles.
+
+    Dataset format:
+
+    * Dataset in single HDF5 file
+    * Json files for band statistics, splits, and tile information
+
+    For additional information, as well as bash scripts to
+    download the data, please refer to the
+    `official repository <https://github.com/vishalned/MMEarth-data?tab=readme-ov-file#data-download>`_.
+
+    If you use this dataset in your research, please cite the following paper:
+
+    * https://arxiv.org/abs/2405.02771
+
+    .. note::
+
+       This dataset requires the following additional library to be installed:
+
+       * `h5py <https://pypi.org/project/h5py/>`_ to load the dataset
+
+    .. versionadded:: 0.6
+    """
+
+    ds_versions: ClassVar[list[str]] = ['MMEarth', 'MMEarth64', 'MMEarth100k']
+
+    filenames: ClassVar[dict[str, str]] = {
+        'MMEarth': 'data_1M_v001',
+        'MMEarth64': 'data_1M_v001_64',
+        'MMEarth100k': 'data_100k_v001',
+    }
+    splits: ClassVar[list[str]] = ['train', 'val', 'test']
+
+    all_modalities: ClassVar[Sequence[str]] = (
```

Comment:
A normal tuple is fine here. Function inputs should be as broad as possible. Function outputs (and class attributes) should be as narrow as possible.